transport-msg.tex

```diff
@@ -1,55 +1,246 @@
 
-The messages transport is designed to work efficiently between VMs on an hypervisor
-based configuration or between heterogeneous systems by using messages to
-reduce the number of context switches between the device and the driver sides
-when each of them is in a different VM. In the same way, this is also reducing
-cross system communications required in an heterogeneous system configuration.
+\subsection{Introduction}
+\label{sec:Virtio Transport Options / Virtio Over Messages / Introduction}
 
-\subsection{Messages Format}\label{sec:Virtio Transport Options / Virtio Over Messages / Messages Format}
+This section defines \textbf{virtio-msg}, a transport mechanism that encapsulates
+virtio operations as discrete message exchanges rather than relying on PCI or
+memory-mapped I/O regions. It separates bus-level functionality (e.g., device
+enumeration, hotplug events) from device-specific operations (e.g., feature
+negotiation, virtqueue setup), ensuring that a single, generic transport layer
+can be reused across multiple bus implementations.
+
+\subsubsection{Purpose}
+
+virtio-msg addresses several key objectives:
 
-The messages have the following properties:
 \begin{itemize}
-\item A message size is 40 bytes.
-\item Most messages are driver/device initialization messages.
-\item Messages are expected to be sent at low frequencies and quick round-trip is
-  not necessary and should not impact global performances (as communication
-  requiring performance should use virtqueues).
+  \item \textbf{Support multiple bus implementations:}
+  Systems may rely on various communication methods such as hypercalls, local
+  IPC, network channels, or device trees for enumerating devices. virtio-msg
+  defines a common transport interface suitable for any of these underlying
+  mechanisms.
+
+  \item \textbf{Reduce per-bus complexity:}
+  Buses can implement a fully message-based workflow (including optional
+  enumeration via \busref{GET_DEVICES} and hotplug via \busref{DEVICE_ADDED}
+  / \busref{DEVICE_REMOVED}), or they can discover and manage devices through
+  alternative means such as platform firmware data. In either case, they
+  forward transport messages to and from each device.
+
+  \item \textbf{Preserve virtio semantics:}
+  The transport leverages standard virtio concepts (features, configuration
+  space, virtqueues), so existing virtio drivers and device backends can
+  integrate smoothly once a device is discovered and registered.
 \end{itemize}
 
-The virtio-MSG messages are encoded as following:
+\subsubsection{High-Level Architecture}
+
+virtio-msg operates around two layers:
+
+\begin{enumerate}
+  \item \textbf{Bus Layer}: A \emph{bus instance} exposes zero or more virtio
+        devices. It can detect available devices through:
+    \begin{itemize}
+      \item Optional message-based queries (\busref{GET_DEVICES}),
+      \item External data sources (e.g., device tree, ACPI tables, hypervisor
```

Comment:
Is this where `FFA_PARTITION_INFO_GET` would fall?

Reply:
When using the FF-A virtio-msg bus the idea is that:
- you use FFA_PARTITION_INFO_GET on the FF-A virtio-msg bus frontend with Virtio Backend UUID to find the endpoints on the system that are hosting a virtio-msg bus backend.
- you then use VIRTIO_MSG_GET_DEVICES to retrieve the list of virtio devices available on an endpoint